wireapp/wire-ios-mono
# fix: no notifications after logout and login again WPB-5365

<table>
<td>
  <a href="https://wearezeta.atlassian.net/browse/WPB-5365" title="WPB-5365" target="_blank"><img alt="Bug" src="https://wearezeta.atlassian.net/rest/api/2/universal_avatar/view/type/issuetype/avatar/10803?size=medium" />WPB-5365</a>  [iOS] No notifications available after logout / login
  </td></table>
  <br />
 






----
#### PR Submission Checklist for internal contributors

- The **PR Title**

- The **PR Description**
----

# What's new in this PR?

### Issues

No notification is shown after logging out and back in again. Also, the user is logged out after receiving a message. Steps to reproduce:

- log in with user A.
- put app in background, receive a message, notification is shown.
- log out, then log back in again with user A.
- put app in background, receive a message, no notification is shown.
- open app, user is logged out.

### Causes

When logging in you get a cookie from the backend. This cookie is stored archived, encrypted, and stored in the keychain. It is also stored in an in-memory cache so we don't need to fetch from the keychain, decrypt, and decode every time we need the cookie. Without this cache, the main app would suffer performance problems.

When the notification extension is loaded because the device receives a push notification, it reads the cookie from the keychain, stores it in the cache, and uses it to request an access token so it can make authenticated requests to the backend.

When the user logs out (in the main app), the cookie is deleted from the cache and the keychain. When the user logs in again, a new cookie is received and stored in the keychain.

If the user receives another push for a new message, then it's likely that the same instance of the notification extension is re-used. Since the user account is the same, the extension will try to fetch the cookie, but will find a hit in the cache. This cached cookie is the old cookie and therefore when an access token is requested it will fail. As a result, we delete the (valid) cookie from the keychain and abort. This is why we don't see any notifications, and when the user opens the main app again, the user is logged out (since there is no cookie in the keychain anymore).

### Solutions

Use the cache only for the main app, don't use the cache for the notification service extension and the share extension because these processes need to make sure they always use current cookie.

### Testing

#### Test Coverage

- unit tests that the cache is used, and not used.

#### How to Test

- log in with user A.
- put app in background, receive a message, notification is shown.
- log out, then log back in again with user A.
- put app in background, receive a message, **assert notification is shown**
- open app, **assert user is not logged out**

### Notes

The problems described here should also affect the share extension, though I didn't test it.


----
#### PR Post Submission Checklist for internal contributors (Optional)

----
#### PR Post Merge Checklist for internal contributors

----
##### References
1. https://sparkbox.com/foundry/semantic_commit_messages
1. https://github.com/wireapp/.github#usage
1. E.g. `feat(conversation-list): Sort conversations by most emojis in the title #SQPIT-764`.
